Remove BC Math Requirement

vinkla opened the pull request:
This pull request will remove the `ext-bcmath` requirement from the `composer.json` file. If you don't have installed BC Math but GMP installed you can't currently install this package. This pull request will solve that issue.

This is still work in progress. The tests should pass and we want to keep our coverage at 100%.

roukmoute commented:
I'm sorry but I don't understand your P.R.
Your commits are not atomic so many of them are strange.

Next, you remove bcmath, but on many methods, you just test and "crach" if bcmaths functions doesn't installed.

In fact I was expecting you have a polyfill for them.

vinkla commented:
> I'm sorry but I don't understand your P.R.

This pull request makes the math extension optional. You will still need to install at least one math extension but we don't force you to always install BC Math if you want to use GMP.

> Your commits are not atomic so many of them are strange.

I've cleaned up the history and squashed the commits.

> Next, you remove bcmath, but on many methods, you just test and "crach" if bcmaths functions doesn't installed.

Yes, since we need to test all scenarios. With BC Math, with GMP and without any math extension installed. If you have any better solution, please let us know.

> In fact I was expecting you have a polyfill for them.

Instead of telling us what you were expecting, we would gladly appreciate any constructive feedback, solution or idea you might have.

vinkla commented:
@ivanakimov when you've time to spare, please let me know what you make of this pull request.

ivanakimov commented:
@vinkla yep will do, haven't caught up yet.

ivanakimov commented:
@vinkla is the user required to have either one or the other library? or are they still optional?

vinkla commented:
> @vinkla is the user required to have either one or the other library? or are they still optional?

Nope, now they can install either `gmp` or `bcmatch` without this library forcing them to install the `bcmatch` extension.

> Edit: because initially numbers larger than a certain amount would simply not be encoded. do you see any scenarios that could break existing prod setups?

This shouldn't break anything in existing setups. The only difference with this change is that we throw an exceptions instead of PHP.

Nyholm commented on tests/AbstractTestCase.php:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @author Vincent Klaiber <hello@vinkla.com>
+ */
+abstract class AbstractTestCase extends TestCase
```
Do we still need this class?

vinkla replied:
True, I'll remove this class.

Nyholm commented on tests/RuntimeExceptionTest.php:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    /**
+     * @expectedException \RuntimeException
```
PHPUnit is moving away for annotations like this. We should consider writing `$this->expectException(\RuntimeException::class);`

vinkla replied:
Didn't know about that. Thanks!

vinkla commented:
Thanks for helping out @Nyholm!
